tests/msc3083_test.go

```diff
@@ -181,6 +184,275 @@ func TestRestrictedRoomsRemoteJoin(t *testing.T) {
+		"user_id": bob.UserID,
+	}
+	res := alice.DoFunc(
+		t,
+		"POST",
+		[]string{"_matrix", "client", "r0", "rooms", room, "invite"},
+		client.WithJSONBody(t, body),
+	)
+	must.MatchResponse(t, res, match.HTTPResponse{
+		StatusCode: 403,
+	})
+
+	// Bob cannot join the room.
+	failJoinRoom(t, bob, room, "hs1", 403)
+
+	// Join the space, joining the room should work, but will bounce through hs2.
+	bob.JoinRoom(t, space, []string{"hs2"})
+	bob.JoinRoom(t, room, []string{"hs1"})
+
+	// Ensure that the join comes down sync on hs2. Note that we want to ensure hs2
+	// accepted the event.
+	charlie.SyncUntilTimelineHas(
+		t,
+		room,
+		func(ev gjson.Result) bool {
+			if ev.Get("type").Str != "m.room.member" || ev.Get("state_key").Str != bob.UserID {
+				return false
+			}
+			must.EqualStr(t, ev.Get("sender").Str, bob.UserID, "Bob should have joined by himself")
+			must.EqualStr(t, ev.Get("content").Get("membership").Str, "join", "Bob failed to join the room")
+
+			return true
+		},
+	)
+
+	// Raise the power level so that users on hs1 can invite people and then leave
+	// the room.
+	state_key := ""
+	charlie.SendEventSynced(t, room, b.Event{
+		Type:     "m.room.power_levels",
+		StateKey: &state_key,
+		Content: map[string]interface{}{
+			"invite": 0,
+			"users": map[string]interface{}{
+				charlie.UserID: 100,
+			},
+		},
+	})
+	charlie.LeaveRoom(t, room)
+
+	// Ensure the events have synced to hs1.
+	alice.SyncUntilTimelineHas(
+		t,
+		room,
+		func(ev gjson.Result) bool {
+			if ev.Get("type").Str != "m.room.member" || ev.Get("state_key").Str != charlie.UserID {
+				return false
+			}
+			must.EqualStr(t, ev.Get("content").Get("membership").Str, "leave", "Charlie failed to leave the room")
+
+			return true
+		},
+	)
+
+	// Have bob leave and rejoin. This should still work even though hs2 isn't in
+	// the room anymore!
+	bob.LeaveRoom(t, room)
+	bob.JoinRoom(t, room, []string{"hs1"})
+}
+
+// A server will request a failover if asked to /make_join and it does not have
+// the appropriate authorisation to complete the request.
+//
+// Setup 3 homeservers:
+// * hs1 creates the space/room.
+// * hs2 joins the room
+// * hs3 attempts to join via hs2 (should fail) and hs3 (should work)
+func TestRestrictedRoomsRemoteJoinFailOver(t *testing.T) {
+	deployment := Deploy(t, b.Blueprint{
+		Name: "federation_three_homeservers",
+		Homeservers: []b.Homeserver{
+			{
+				Name: "hs1",
+				Users: []b.User{
+					{
+						Localpart:   "alice",
+						DisplayName: "Alice",
+					},
+				},
+			},
+			{
+				Name: "hs2",
+				Users: []b.User{
+					{
+						Localpart:   "bob",
+						DisplayName: "Bob",
+					},
+				},
+			},
+			{
+				Name: "hs3",
+				Users: []b.User{
+					{
+						Localpart:   "charlie",
+						DisplayName: "Charlie",
+					},
+				},
+			},
+		},
+	})
+	defer deployment.Destroy(t)
+
+	// Setup the user, space, and restricted room.
+	alice, space, room := setupRestrictedRoom(t, deployment)
```

Comment:
Why don't you use `setupRestrictedRoom` in the previous test?

Reply:
The joiners need to be on the same server while the creator needs to be on a separate server. Unfortunately `BlueprintFederationTwoLocalOneRemote` has alice and bob on one server and charlie on another.